mlflow/utils/autologging_utils.py

```diff
@@ -260,3 +292,409 @@ def batch_metrics_logger(run_id):
+            # This is important because several autologging patch implementations inspect
+            # the signature of the `original` argument during execution
+            call_original = _update_wrapper_extended(call_original, original)
+
+            if patch_is_class:
+                patch_function.call(call_original, *args, **kwargs)
+            else:
+                patch_function(call_original, *args, **kwargs)
+
+        except Exception as e:  # pylint: disable=broad-except
+            # Exceptions thrown during execution of the original function should be propagated
+            # to the caller. Additionally, exceptions encountered during test mode should be
+            # reraised to detect bugs in autologging implementations
+            if failed_during_original or _is_testing():
+                raise
+
+            _logger.warning(
+                "Encountered unexpected error during %s autologging: %s", autologging_integration, e
+            )
+
+        if original_has_been_called:
+            return original_result
+        else:
+            return original(*args, **kwargs)
+
+    wrap_patch(destination, function_name, safe_patch_function)
+
+
+def _validate_args(
+    user_call_args, user_call_kwargs, autologging_call_args, autologging_call_kwargs
+):
+    """
+    Used for testing purposes to verify that, when a patched ML function calls its underlying
+    / original ML function, the following properties are satisfied:
+
+        - All arguments supplied to the patched ML function are forwarded to the
+          original ML function
+        - Any additional arguments supplied to the original function are exception safe (i.e.
+          they are either functions decorated with the `@exception_safe_function` decorator
+          or classes / instances of classes with type `ExceptionSafeClass`
+    """
+
+    def _validate_new_input(inp):
+        """
+        Validates a new input (arg or kwarg) introduced to the underlying / original ML function
+        call during the execution of a patched ML function. The new input is valid if:
+
+            - The new input is a function that has been decorated with `exception_safe_function`
+            - OR the new input is a class with the `ExceptionSafeClass` metaclass
+            - OR the new input is a list and each of its elements is valid according to the
+              these criteria
+        """
+        if type(inp) == list:
+            for item in inp:
+                _validate_new_input(item)
+        elif callable(inp):
+            assert getattr(inp, _ATTRIBUTE_EXCEPTION_SAFE, False), (
+                "New function argument '{}' passed to original function is not exception-safe."
+                " Please decorate the function with `exception_safe_function`.".format(inp)
+            )
+        elif inspect.isclass(type(inp)):
+            assert type(inp.__class__) == ExceptionSafeClass, (
```

Comment:
```suggestion
            assert inp.__class__ == ExceptionSafeClass, (
```

Is `type` necessary here?

```python
>>> class Foo:
        pass
>>> f = Foo()

>>> f.__class__
<class '__main__.Foo'>

>>> type(f.__class__)
<class 'type'>
```

Reply:
Just never mind. `ExceptionSafeClass` is used as a meta class so we need `type`